test/listen-dial.spec.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+const nativeUTP = require('utp-native')
+const pull = require('pull-stream')
+const multiaddr = require('multiaddr')
+
+const UTP = require('../src')
+const isCI = process.env.CI
+
+describe('listen', () => {
+  let utp
+
+  function ma (port) {
+    const base = '/ip4/127.0.0.1/udp/'
+    return multiaddr(`${base}${port}/utp`)
+  }
+
+  beforeEach(() => {
+    utp = new UTP()
+  })
+
+  it('close listener with connections, through timeout', function (done) {
+    this.timeout(20 * 1000)
+
+    const listener = utp.createListener((conn) => {
+      pull(conn, conn)
+    })
+
+    listener.on('connection', () => {
+      // Testing
+      console.log('incomming connection')
+    })
+
+    const addr = ma(12000)
+    const connectOptions = addr.toOptions()
+
+    listener.listen(addr, () => {
+      const socket1 = nativeUTP.connect(connectOptions.port, connectOptions.host)
+      const socket2 = nativeUTP.connect(connectOptions.port, connectOptions.host)
+
+      socket1.write('Some data that is never handled')
+      socket1.end()
+
+      // TODO Unfortunately utp has no notion of gracious socket closing
+      // This feature needs to be shimmed on top to make it a proper libp2p
+      // transport
```

Comment:
**This continues to be the crux to support utp**. libp2p currently expect gracious connection closing to be supported. We either remove this requirement or add a shim to libp2p-utp.

Reply:
Update: Calling `socket.end()` _might_ fix it... investigating
https://github.com/mafintosh/utp-native/blob/master/src/socket_wrap.cc#L183-L186

Reply:
It _does_ send fin, but for some reason only from 1 side
![screenshot](https://user-images.githubusercontent.com/7735145/39177968-df868758-47b0-11e8-992c-8ce57b009af0.png)
[utp_traffic.pcapng.gz](https://github.com/libp2p/js-libp2p-utp/files/1942018/utp_traffic.pcapng.gz)

Reply:
Another update: `server.close()` never emits the close event.

Reply:
```
test/net-api.js:261:    socket.end() // utp does not support half open
```

Just found this.